cmake/modules/VersionInfo.cmake

```diff
@@ -9,33 +9,62 @@ IF(GIT_FOUND AND NOT FORCE_VERSION)
 		OUTPUT_STRIP_TRAILING_WHITESPACE)
 	STRING(REPLACE "-" ";" TAG_LIST "${GIT_TAG}")
 	LIST(LENGTH TAG_LIST TAG_LIST_LENGTH)
-	IF(TAG_LIST_LENGTH EQUAL 1)
-		# Stable build, FORCE_VERSION=x.x.x
-		LIST(GET TAG_LIST 0 FORCE_VERSION)
-		STRING(REPLACE "v" "" FORCE_VERSION "${FORCE_VERSION}")
-	ELSEIF(TAG_LIST_LENGTH EQUAL 3)
-		# Non-stable build, FORCE_VERSION=x.x.x-hash
-		LIST(GET TAG_LIST 0 FORCE_VERSION)
-		LIST(GET TAG_LIST 2 COMMIT_HASH)
-		STRING(REPLACE "v" "" FORCE_VERSION "${FORCE_VERSION}")
-		STRING(REPLACE "g" "" COMMIT_HASH "${COMMIT_HASH}")
-		SET(FORCE_VERSION "${FORCE_VERSION}-${COMMIT_HASH}")
+	LIST(GET TAG_LIST 0 FORCE_VERSION)
+	STRING(REPLACE "v" "" FORCE_VERSION "${FORCE_VERSION}")
+	IF(TAG_LIST_LENGTH EQUAL 3)
+		# Non-tagged version
+		LIST(GET TAG_LIST 1 EXTRA_COMMITS)
+		STRING(REPLACE "." ";" VERSION_LIST "${FORCE_VERSION}")
+		LIST(LENGTH VERSION_LIST VERSION_LENGTH)
+		LIST(GET VERSION_LIST 0 VERSION_MAJOR)
+		LIST(GET VERSION_LIST 1 VERSION_MINOR)
+		LIST(GET VERSION_LIST 2 VERSION_BUILD)
+		SET(VERSION_REVISION 0)
+		IF(VERSION_LENGTH GREATER 3)
+			LIST(GET VERSION_LIST 3 VERSION_REVISION)
+		ENDIF()
+
+		EXECUTE_PROCESS(
+			COMMAND "${GIT_EXECUTABLE}" rev-parse --abbrev-ref HEAD
+			OUTPUT_VARIABLE BRANCH_NAME
+			WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}"
+			OUTPUT_STRIP_TRAILING_WHITESPACE)
+		STRING(REGEX MATCH "^stable-${VERSION_MAJOR}\\.${VERSION_MINOR}"
+			STABLE_BRANCH "${BRANCH_NAME}")
+		IF(STABLE_BRANCH)
+			# Stable update, FORCE_VERSION=x.x.x.(x + commits)
+			MATH(EXPR VERSION_REVISION
+				"${VERSION_REVISION} + ${EXTRA_COMMITS}")
+			SET(FORCE_VERSION
+	"${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_BUILD}.${VERSION_REVISION}"
+			)
+		ELSE()
+			# Development, FORCE_VERSION=x.x.(x + commits)
+			MATH(EXPR VERSION_BUILD
+				"${VERSION_BUILD} + ${EXTRA_COMMITS}")
+			SET(FORCE_VERSION
+			"${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_BUILD}"
+			)
+		ENDIF()
 	ENDIF()
 ENDIF()
 
 IF(FORCE_VERSION STREQUAL "internal")
 	# Use release info from /CMakeLists.txt
 ELSEIF(FORCE_VERSION)
 	STRING(REPLACE "." ";" VERSION_LIST "${FORCE_VERSION}")
-	STRING(REPLACE "-" ";" VERSION_LIST "${VERSION_LIST}")
 	LIST(LENGTH VERSION_LIST VERSION_LENGTH)
 	LIST(GET VERSION_LIST 0 VERSION_MAJOR)
 	LIST(GET VERSION_LIST 1 VERSION_MINOR)
-	LIST(GET VERSION_LIST 2 VERSION_PATCH)
+	LIST(GET VERSION_LIST 2 VERSION_BUILD)
 	IF(VERSION_LENGTH GREATER 3)
-		LIST(GET VERSION_LIST 3 VERSION_SUFFIX)
+		LIST(GET VERSION_LIST 3 VERSION_REVISION)
 	ENDIF()
 	SET(VERSION             "${FORCE_VERSION}")
+ELSEIF(GIT_FOUND)
+	MESSAGE(
```

Comment:
I do work with 80-char terminals for efficiency reasons:

* I have a lot of terminals open, it is a fast way to navigate through code, and I often need to compare outputs side by side; two terminals take up my screen width.
* 80-char lines avoid horizontal scrolling. Vertical scrolling is friendlier; I have the mouse wheel, the touchpad and page up/down keys.
* The flow is faster; when I scroll to the right, then I have to scroll back to the left.
* When looking at the right part of one line, I can only pay attention to that line. I lose the context of other lines.
* With `nano`, one editor I use, it is difficult to edit the part immediately after the 80-char limit.
* When selecting long lines, I have to be careful not to move the mouse over or under the line, because I lose the selection and the position. I usually select long lines with the keyboard, and selecting with the keyboard is slower.

So I know 80-char lines are a friendly thing. I can work with longer lines, but they are harder to maintain.